File: Data Structures/#08.Heaps & Priority Queue Lab/Java/src/main/java/Heap.java
public class Heap<T extends Comparable<T>> {

    public static <T extends Comparable<T>> void sort(T[] array) {

        if(array.length == 0){
            return;
        }

        for (int i = array.length / 2; i >= 0; i--) {
            heapifyDown(array, i);
        }

        Object[] sortedArr = new Object[array.length];
        for (int i = array.length - 1; i >= 0; i--) {
            swap(array, 0, i);
            sortedArr[i] = array[i];
            array[i] = null;
            heapifyDown(array, 0);
        }

        for (int i = 0; i < array.length; i++) {
            array[i] = (T) sortedArr[i];
        }
    }

    private static <T extends Comparable<T>> void heapifyDown(T[] array, int elementIndex) {
        T element = array[elementIndex];

        while (true) {
            boolean changes = false;
            T leftChild = leftChild(array, elementIndex);
            T rightChild = rightChild(array, elementIndex);

            if (leftChild != null)
                if (isHigherThan(element, leftChild)) {
                    int leftChildIndex = ((2 * elementIndex) + 1);

                    swap(array, elementIndex, leftChildIndex);

                    elementIndex = leftChildIndex;
                    leftChild = leftChild(array, elementIndex);
                    rightChild = rightChild(array, elementIndex);
                    changes = true;
                }

            if (rightChild != null)
                if (isHigherThan(element, rightChild)) {
                    int rightChildIndex = ((2 * elementIndex) + 2);

                    swap(array, elementIndex, rightChildIndex);

                    elementIndex = rightChildIndex;
                    leftChild = leftChild(array, elementIndex);
                    rightChild = rightChild(array, elementIndex);
                    changes = true;
                }

            if (!changes) {
                break;
            }
        }

    }

    private static <T extends Comparable<T>> void swap(T[] array, int index1, int index2) {
        T index1Element = array[index1];
        T index2Element = array[index2];

        array[index1] = index2Element;
        array[index2] = index1Element;
    }

    private static <T extends Comparable<T>> boolean isHigherThan(T element1, T element2) {
        return element1.compareTo(element2) < 0;
    }

    private static <T extends Comparable<T>> T leftChild(T[] array, int elementIndex) {
        int index = ((2 * elementIndex) + 1);
        return index > array.length - 1 ? null : array[index];
    }

    private static <T extends Comparable<T>> T rightChild(T[] array, int elementIndex) {
        int index = ((2 * elementIndex) + 2);
        return index > array.length - 1 ? null : array[index];
    }


}

File: Data Structures/#08.Heaps & Priority Queue Lab/C#/BinaryHeap/Heap.cs
using System;
using System.Collections.Generic;

public static class Heap<T> where T : IComparable<T>
{
    public static void Sort(T[] arr)
    {
        //Creates max heap
        for (int i = 0; i < arr.Length; i++)
        {
            HeapifyUp(i, arr);
        }

        for (int i = arr.Length - 1; i >= 0; i--)
        {
            Swap(0, i, arr);
            HeapifyDown(0, arr, i);
        }
        Console.WriteLine();
    }

    private static void HeapifyDown(int elementIndex, T[] heap, int border)
    {

        while (elementIndex < border / 2)
        {
            T element = heap[elementIndex];
            int child = 2 * elementIndex + 1;

            if (child + 1 < border && IsLess(heap,child,child + 1))
            {
                child++;
            }

            if (IsLess(heap, child, elementIndex))
            {
                break;
            }

            Swap(elementIndex,child,heap);
            elementIndex = child;
        }
    }

    private static bool IsLess(T[] heap, int element1Index, int element2Index)
    {
        return heap[element1Index].CompareTo(heap[element2Index]) < 0;
    }

    private static void HeapifyUp(int elementIndex, T[] heap)
    {
        T element = heap[elementIndex];
        int parentIndex = (elementIndex - 1) / 2;
        T parent = heap[parentIndex];

        while (element.CompareTo(parent) > 0)
        {
            Swap(elementIndex, parentIndex, heap);

            //update values
            elementIndex = parentIndex;
            parentIndex = (elementIndex - 1) / 2;
            element = heap[elementIndex];
            parent = heap[parentIndex];
        }
    }

    private static void Swap(int index1, int index2, T[] heap)
    {
        T index1Element = heap[index1];

        heap[index1] = heap[index2];
        heap[index2] = index1Element;
    }

}
